src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -2,23 +2,220 @@
+                        // does not perform as well for smaller arrays.
+                        var result = new T[finalLength];
+                        for (int i = 0; i < result.Length; i++) result[i] = buffer[i];
+                        return result;
+                    }
+
+                    // We called Current last, but the loop expects MoveNext to be called
+                    // before it is entered. So we need to call it again here.
+                    if (!en.MoveNext())
+                    {
+                        // The enumerable fits perfectly into the buffer, so no resizing to do
+                        return buffer;
+                    }
+                }
+                while (buffer.Length != BufferListThreshold);
+
+                Debug.Assert(buffer.Length == BufferListThreshold); // The only way we should have exited the loop was if the loop condition was false
+
+                // Since this path is only going to get called for lazy enumerables of length > 32, do not
+                // force the jit to compile lots of code it will never use (esp. since this method is generic).
+                return LazyToArrayUsingBufferList(en, buffer);
+            }
+        }
+
+        private static T[] LazyToArrayUsingBufferList<T>(IEnumerator<T> en, T[] first)
+        {
+            // NOTE: You are expected to call MoveNext after the last call to Current
+            // before passing in the enumerator to this method.
+
+            Debug.Assert(en != null);
+            Debug.Assert(first != null);
+
+            // Instead of further resizing the array, we're going to maintain a list
+            // of buffers. Each time we can't fit the sequence into a buffer, we're
+            // going to store the buffer in this list, create a new one 2x the size,
+            // and continue reading in the sequence.
+            // The data is no longer going to be contiguous in memory, but that doesn't
+            // matter since even if we were using the resize-every-time approach, we'd
+            // still have to resize again at the end to make the array exactly the right size.
+
+            // Here's a visualization showing what first, buffers, and current
+            // might look like for a 200-length enumerable:
+            //
+            // first: [items 0-31]
+            //
+            // buffers:
+            // [0]: [items 32-63]
+            // [1]: [items 64-127]
+            //
+            // current: [items 128-199], [slots 200-255 empty]
+
+            // The up-front allocation for the list will not be that much since
+            // the backing store will be Array.Empty<T>() if you do not pass in a
+            // capacity. It only starts allocating arrays when we add the first item.
+
+            var buffers = new List<T[]>(); // list of previous buffers
+            var current = new T[first.Length]; // the current buffer we're reading the sequence into
```

Comment:
@NSil That seems like it would increase allocations for e.g. lengths 33-64, since before we would allocate a `T[32]` and a `T[32]`, and now we would allocate a `T[32]` and `T[64]`. What would be an example in which is decreases allocations/copying?